lib/core/utils/preload-cssom.js

```diff
@@ -0,0 +1,231 @@
+ * Construct an extended object of cssom stylesheet with added attribtues for isExternal and shadowId
+ * @method getCssomSheet
+ * @private
+ * @param {CSSStyleSheet} sheet stylesheet
+ * @param {Boolean} isExternal flag to specify if the stylesheet was fetched via xhr
+ * @param {String} shadowId (optional) string representing shadowId if style/ sheet was constructed from shadowDOM assets
+ */
+function getCssomSheet(sheet, isExternal, shadowId) {
+	const out = { isExternal, sheet };
+	if (shadowId) {
+		out.shadowId = shadowId;
+	}
+	return out;
+}
+
+/**
+ * Returns a then(able) queue of CSSStyleSheet(s)
+ * @param {Object} ownerDocument document object to be inspected for stylesheets
+ * @param {number} timeout on network request for stylesheet that need to be externally fetched
+ * @param {Function} convertTextToStylesheetFn a utility function to generate a style sheet from text
+ * @return {Object} queue
+ * @private
+ */
+function loadCssom({ root, shadowId }, timeout, convertTextToStylesheetFn) {
+	/**
+	 * Make an axios get request to fetch a given resource and resolve
+	 * @method getExternalStylesheet
+	 * @private
+	 * @param {Object} param an object with properties to configure the external XHR
+	 * @property {Object} param.resolve resolve callback on queue
+	 * @property {Object} param.reject reject callback on queue
+	 * @property {String} param.url string representing the url of the resource to load
+	 * @property {Number} param.timeout timeout to about network call
+	 */
+	function getExternalStylesheet({ resolve, reject, url }) {
+		axe.imports
+			.axios({
+				method: 'get',
+				url,
+				timeout
+			})
+			.then(({ data }) => {
+				const sheet = convertTextToStylesheetFn({
+					data,
+					isExternal: true,
+					shadowId
+				});
+				resolve(sheet);
+			})
+			.catch(reject);
+	}
+
+	const q = axe.utils.queue();
+	// iterate to decipher multi-level nested sheets if any (this is essential to retrieve styles from shadowDOM)
+	Array.from(root.styleSheets).forEach(sheet => {
+		// ignore disabled sheets
+		if (sheet.disabled) {
+			return;
+		}
+		// attempt to retrieve cssRules, or for external sheets make a XMLHttpRequest
+		try {
+			// accessing .cssRules throws for external (cross-domain) sheets, which is handled in the catch
+			const cssRules = sheet.cssRules;
+			// read all css rules in the sheet
+			const rules = Array.from(cssRules);
+			// filter rules to assess if any external rules by way of @import or link exists on a nested level
+			const externalRules = rules.filter(r => r.href);
+			// if no external rules then return an expanded version of
+			if (!externalRules.length) {
+				q.defer(resolve => resolve(getCssomSheet(sheet, false, shadowId)));
+				return;
+			}
+
+			// there are external rules, best to filter non-external rules and create a new stylesheet to avoid duplication
+			const nonExternalRules = rules.filter(rule => !rule.href);
+			// concat all cssText into a string for non external rules
```

Comment:
I'm having a very hard time understanding what's happening here. As best I can tell you're trying to turn same origin sheets and style elements with import statements into one long stylesheet... that doesn't seem like a good idea to me. This means we lose all context of how why and where a stylesheet got included. It also ignores that `@import` can have its own media value, and that they can be recursive. We also shouldn't ignore `@import` on cross-origin sheets.

Please be sure to include tests that covers all of these things.

Reply:
Note: Paired on the above, refactored as discussed.